### ISSUE TITLE
identity_op: fails to apply fix for reference

### ISSUE THREAD

**Issue by FineFindus** [rust-lang/rust-clippy#12050]
### Summary

When attempting to fix the `identity_op` lint for variables that are references, such as `&i32`, the lint fails to apply.

This issue is similar to https://github.com/rust-lang/rust-clippy/issues/9904, which is correctly resolved by removing the reference.

### Reproducer

I tried this code:

```rust
fn main() {
    let x = &0i32;
    let _: i32 = x + 0;
}
```

I expected to see this happen:
When running `cargo clippy --fix` the variable should be dereferenced.
```rust
fn main() {
    let x = &0i32;
    let _: i32 = *x;
}
```

Instead, this happened:

```
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_bug_report`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
warning: failed to connect to jobserver from environment variable `CARGO_MAKEFLAGS="-j --jobserver-fds=8,9 --jobserver-auth=8,9"`: cannot open file descriptor 8 from the jobserver environment variable value: Bad file descriptor (os error 9)
  |
  = note: the build environment is likely misconfigured

error[E0308]: mismatched types
 --> src/main.rs:3:18
  |
3 |     let _: i32 = x;
  |            ---   ^ expected `i32`, found `&i32`
  |            |
  |            expected due to this
  |
help: consider dereferencing the borrow
  |
3 |     let _: i32 = *x;
  |                  +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: this operation has no effect
 --> src/main.rs:3:18
  |
3 |     let _: i32 = x + 0;
  |                  ^^^^^ help: consider reducing it to: `x`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_op
  = note: `#[warn(clippy::identity_op)]` on by default

warning: `clippy-bug-report` (bin "clippy-bug-report") generated 1 warning (run `cargo clippy --fix --bin "clippy-bug-report"` to apply 1 suggestion)
warning: `clippy-bug-report` (bin "clippy-bug-report" test) generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
```

### Version

```text
rustc 1.77.0-nightly (fb5ed726f 2023-12-28)
binary: rustc
commit-hash: fb5ed726f72c6d16c788517c60ec00d4564b9348
commit-date: 2023-12-28
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

**Comment by PartiallyTyped** [comment]
Okay, so what's happening is that when a ref is implicitly deref'd / coerced into the value, the lint doesn't take it into consideration.

This fails, x is ref and gets deref'd.
```
fn main() {
    let x = &0i32;
    let _: i32 = x + 0;
}
```

This gets fixed correctly.
```
fn main() {
    let x = &0i32;
    let _: i32 = *x + 0;
}
```
